---
title: "Selective sweep test expected heterozygosity"
output: html_document
date: "2024-04-09"
editor_options: 
  markdown: 
    wrap: 72
---

# 0: Preparation

## Defining the input and output directories

```{r}
# Clean the working environment
rm(list = ls())

knitr::opts_chunk$set(echo = TRUE)

# Set the path to your GitHub folder
YOUR_GITHUB_ROOT_DIRECTORY <- "C:/Users/jonat/GitHub"

# Defining the relative path in the repository
repository_path <- "Computational-modelling-of-genomic-inbreeding-and-roh-islands-in-extremely-small-populations"


####################################  
# Defining the input files
#################################### 



ROH_hotspot_allele_freq_dir <- file.path(YOUR_GITHUB_ROOT_DIRECTORY, repository_path,"results/ROH-Hotspots/empirical/german_shepherd/gosling_plots/hotspots_allele_freq")

neutral_model_allele_freq_dir <- file.path(YOUR_GITHUB_ROOT_DIRECTORY, repository_path,"results/PLINK/simulated/allele_freq")


####################################  
# Defining the working directory
#################################### 

output_dir_sweep_test <- file.path(YOUR_GITHUB_ROOT_DIRECTORY, repository_path,"results/Sweep_test/german_shepherd")


if (!dir.exists(output_dir_sweep_test)) {
  # Create the working directory if it doesn't exist
  dir.create(output_dir_sweep_test,recursive = TRUE)
}

# Set the working directory for notebook chunks
knitr::opts_knit$set(root.dir = output_dir_sweep_test)

```

## Loading libraries

```{r library()}
library(knitr)
```

# 1: Loading the frequency files

## 1.1: Empirical ROH-Hotspots - Loading the allele frequencies

```{r}
# Set the working directory to the directory with allele frequencies for the different ROH-hotspots
setwd(ROH_hotspot_allele_freq_dir)

# Get a list of all .bed files in the directory
bed_files <- list.files(pattern = "\\.bed$")

# Create an empty list to store information for the different ROH-hotspots in
empirical_hotspot_tables <- list()

# Loop through each .bed file (ROH-hotspot allele frequency window-file)
for (file in bed_files) {
  # Extract chromosome number and window number from file name
  chromosome <- sub(".*chr([0-9]+)_.*", "\\1", file)
  window <- sub(".*window_([0-9]+)_.*", "\\1", file)
  
  # Create table name
  table_name <- paste("Hotspot_chr", chromosome, "_window_", window, "_allele_freq", sep = "")
  
  # Read the header line
  con <- file(file, "r")
  header <- readLines(con, n = 1)
  close(con)
  
  # Remove "#" from the header and split it into column names
  column_names <- sub("#", "", header)
  column_names <- strsplit(column_names, "\t")[[1]]
  
  # Read the .bed file into a data frame, skipping commented lines
  bed_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  
  # Create a list with table name and corresponding data frame
  table_info <- list(name = table_name, data = bed_data)
  
  # Append the table info to the list
  empirical_hotspot_tables <- c(empirical_hotspot_tables, list(table_info))
}

# Print the list of empirical_hotspot_tables
#print(empirical_hotspot_tables)


```

## 1.2: Neutral model - Loading the allele frequencies

```{r}
# Set the working directory to the directory containing .frq files
setwd(neutral_model_allele_freq_dir)

# Get a list of all .frq files in the directory
frq_files <- list.files(pattern = "\\.frq$")

# Initialize an empty list to store neutral model tables
neutral_model_tables <- list()

# Loop through each .frq file
for (file in frq_files) {
 
  # Extract the table name from the file name
  table_name <- gsub(".frq$", "", file)
  
  # Read the .frq file into a data frame
  frq_data <- read.table(file, header = TRUE)
  
  # Create a list with table name and corresponding data frame
  table_info <- list(name = table_name, data = frq_data)
  
  # Append the table info to the list
  neutral_model_tables <- c(neutral_model_tables, list(table_info))
}


# Print the list of neutral_model_tables
#print(neutral_model_tables)

```

# 2: Computing expected heterozygosity

```{r}
calculate_expected_heterozygosity <- function(p) {
  # Calculate expected heterozygosity
  heterozygosity <- 2 * p * (1 - p)
  
  # Return the result
  return(heterozygosity)
}

```

## 2.1: Empirical ROH-Hotspots

```{r}

for (i in seq_along(empirical_hotspot_tables)) {
  # Extract table name and data frame
  table_info <- empirical_hotspot_tables[[i]]
  table_data <- table_info$data
  
  # Calculate expected heterozygosity for each window
  window_H_e_list <- c()  # Initialize an empty list to store expected heterozygosity values
  
  # Loop through each row in the table data
  for (j in 1:nrow(table_data)) {
    # Extract the MAF (p value)
    p <- table_data[j, "MAF"]
    
    # Calculate expected heterozygosity using the provided function
    expected_heterozygosity <- calculate_expected_heterozygosity(p)
    
    # Append the calculated expected heterozygosity to the list
    window_H_e_list <- c(window_H_e_list, expected_heterozygosity)
  }
  
  # Assign the calculated expected heterozygosity list to the H_e attribute of table_info
  empirical_hotspot_tables[[i]]$H_e_list <- window_H_e_list
  empirical_hotspot_tables[[i]]$Average_H_e <- mean(empirical_hotspot_tables[[i]]$H_e_list)
}

#View(empirical_hotspot_tables)

```

## 2.2: Neutral model

```{r}
for (i in seq_along(neutral_model_tables)) {
  # Extract table name and data frame
  table_info <- neutral_model_tables[[i]]
  table_data <- table_info$data
  
  # Calculate expected heterozygosity for each window
  window_H_e_list <- c()  # Initialize an empty list to store expected heterozygosity values
  
  # Loop through each row in the table data
  for (j in 1:nrow(table_data)) {
    # Extract the MAF (p value)
    p <- table_data[j, "MAF"]
    
    # Calculate expected heterozygosity using the provided function
    expected_heterozygosity <- calculate_expected_heterozygosity(p)
    
    # Append the calculated expected heterozygosity to the list
    window_H_e_list <- c(window_H_e_list, expected_heterozygosity)
  }
  
  # Assign the calculated expected heterozygosity list to the H_e attribute of table_info
  neutral_model_tables[[i]]$H_e_list <- window_H_e_list
  neutral_model_tables[[i]]$Average_H_e <- mean(neutral_model_tables[[i]]$H_e_list)
}

#View(neutral_model_tables)

```

# 3: Sweep test

## 3.1: Neutral model - H_e distribution

```{r}
# Extract Average_H_e values from all tables in neutral_model_tables
sim_average_H_e_values <- sapply(neutral_model_tables, function(table) table[["Average_H_e"]])

# Determine the number of bins you want
num_bins <- 20  

# Create a histogram with custom breaks
hist(sim_average_H_e_values, breaks = num_bins, main = "Distribution of Average H_e Values", xlab = "H_e")

# Calculate the 5th percentile
percentile_5 <- quantile(sim_average_H_e_values, 0.05)

# Add a vertical line at the 5th percentile
abline(v = percentile_5, col = "red", lty = 2)

# Add a legend
legend("topright", legend = paste("5th percentile =", round(percentile_5, 4)), col = "red", lty = 2)
```

## 3.2: ROH-Hotspots Selection testing

Selection testing for the different ROH-Hotspot windows.

ROH-Hotspots with an average H_e lower than **or** equal to (\<=) the
5th percentile of the H_e-distribution for the simulated neutral model,
**are ROH-Hotspots deemed as being under selection**.

```{r}
setwd(output_dir_sweep_test)

# Defining the dataframe that will store the selection testing results
ROH_hotspots_selection_testing <- data.frame(
  Name = character(),  
  Under_selection = character(),
  Average_H_e = numeric() 
)


for (i in seq_along(empirical_hotspot_tables)) {
  # Check if Average_H_e is <= percentile_5 for the current table (ROH-hotspot)
  if (empirical_hotspot_tables[[i]][["Average_H_e"]] <= percentile_5) {
    # # If under selection, append the name and Average_H_e value with "Yes"
    ROH_hotspots_selection_testing <- rbind(ROH_hotspots_selection_testing, data.frame(
      Name = empirical_hotspot_tables[[i]][["name"]],
      Under_selection = "Yes",  # Marking as "Yes" if under selection
      Average_H_e = empirical_hotspot_tables[[i]][["Average_H_e"]]
    ))
  } else {
    # If not under selection, append the name and Average_H_e value with "No" 
    ROH_hotspots_selection_testing <- rbind(ROH_hotspots_selection_testing, data.frame(
      Name = empirical_hotspot_tables[[i]][["name"]],
      Under_selection = "No",  # Marking as "No" if not under selection
      Average_H_e = empirical_hotspot_tables[[i]][["Average_H_e"]]
    ))
  }
}

# Sort the data frame by Average_H_e values in ascending order
ROH_hotspots_selection_testing <- ROH_hotspots_selection_testing[order(ROH_hotspots_selection_testing$Average_H_e), ]

# Plot histogram of hotspot Average_H_e values
hist(ROH_hotspots_selection_testing$Average_H_e, breaks = 20, main = "ROH-Hotspots Distribution of Average H_e Values", xlab = "H_e")

# Add a vertical line at the 5th percentile
abline(v = percentile_5, col = "red", lty = 2)

# Add legend for the 5th percentile line
legend("topright", legend = paste("5th percentile =", round(percentile_5, 4)), col = "red", lty = 2)


cat("ROH-hotspot selection testing results:\n")
kable(ROH_hotspots_selection_testing, row.names = FALSE)

# Saving the results in a .tsv-file
write.table(ROH_hotspots_selection_testing, "ROH_hotspots_selection_testing_results.tsv", sep = "\t", row.names = FALSE,quote = FALSE)



```
