---
output: 
  html_document: 
    toc: true
---

# 0: Preparation
Defining the input and output files

```{r setup,echo = FALSE}

#echo = FALSE # Hides the code.
#results = 'hide' # Hides the output of the code.
#message = FALSE # Hides any messages generated by the code.
#warning = FALSE # Hides any warnings generated by the code.

# Clean the working environment
rm(list = ls())


empirical_species <- "Labrador Retriever"
document_title <- paste(empirical_species,"Hyperoptimization Neutral Model")

document_sub <- paste("No MAF-based pruning used")


knitr::opts_chunk$set(echo = TRUE)





############################################ 
# Input Parameters 
############################################ 
empirical_dog_breed <- "labrador_retriever"

# Parameters used for displaying the result
ROH_frequency_decimals <- 1
H_e_values_decimals <- 5
F_ROH_values_decimals <- 3
decimals_output_values <- 5


# chr <- Sys.getenv("chr_simulated")
chr <- "chr3" 

# Ne_burn_in <- as.numeric(Sys.getenv("Ne_burn_in"))
Ne_burn_in <- 250

# nInd_fo <- as.numeric(Sys.getenv("nInd_founder_population"))
nInd_fo <- 50 

# Inbred_fo <- Sys.getenv("Inbred_ancestral_population")
Inbred_fo <- TRUE

# N_e_bottleneck <- as.numeric(Sys.getenv("N_e_bottleneck"))
N_e_bottleneck <- 30 # [30,40,50,60,70]
# n_generations_bottleneck <- as.numeric(Sys.getenv("n_generations_bottleneck"))
n_generations_bottleneck <- 5


# n_gen_breed <- as.numeric(Sys.getenv("n_simulated_generations_breed_formation"))
n_gen_breed <- 30 # [40,45,50,55,60,65,70]

# n_breed <- as.numeric(Sys.getenv("n_individuals_breed_formation"))
n_breed <- 375 # [40-70]


# snp_chip <- Sys.getenv("reference_population_for_snp_chip")
snp_chip <- "last_breed_formation_generation" 

# Introduce_mutations <- Sys.getenv("Introduce_mutations")
Introduce_mutations <- TRUE



####################################  
# Defining the working directory
#################################### 

# Set the path to your GitHub folder
YOUR_GITHUB_ROOT_DIRECTORY <- "C:/Users/jonat/GitHub"

# Defining the relative path in the repository
repository_path <- file.path(YOUR_GITHUB_ROOT_DIRECTORY,"Computational-modelling-of-genomic-inbreeding-and-roh-islands-in-extremely-small-populations")


####################################  
# Defining the input file
#################################### 
# results_dir <- file.path(repository_path,"results_50")
# results_dir <- file.path(repository_path,"results_lab")
results_dir <- file.path(repository_path,"results_HO")


plink_ROH_dir <- file.path(results_dir,"PLINK//ROH")
# expected_heterozygosity_dir <- file.path(results_dir,"expected_heterozygosity")
expected_heterozygosity_dir <- file.path(results_dir,"expected_heterozygosity_No_MAF")

ROH_hotspots_dir <- file.path(results_dir,"ROH-Hotspots")

# Selection_strength_test_dir <- file.path(ROH_hotspots_dir,"selection_strength_test")
Selection_strength_test_dir <- file.path(ROH_hotspots_dir,"selection_strength_test_No_MAF")

# Sweep_test_dir <- file.path(ROH_hotspots_dir,"sweep_test")
Sweep_test_dir <- file.path(ROH_hotspots_dir,"sweep_test_No_MAF")



############### 
## Empirical ###
###############

### ROH hotspots ###
Empirical_data_ROH_hotspots_dir  <- file.path(ROH_hotspots_dir,"empirical",empirical_dog_breed)
Empirical_data_autosome_ROH_freq_dir <- file.path(Empirical_data_ROH_hotspots_dir,"Gosling_plots/autosome_roh_freq")
### Inbreeding coefficient ###
Empirical_data_F_ROH_dir  <- file.path(plink_ROH_dir,"empirical",empirical_dog_breed,"F_ROH")

### Expected Heterozygosity distribution ###
Empirical_data_H_e_dir <- file.path(expected_heterozygosity_dir,"empirical",empirical_dog_breed)

############### 
## Simulated ###
###############

### ROH hotspots ###
Neutral_model_ROH_hotspots_dir  <- file.path(ROH_hotspots_dir,"simulated/neutral")
Neutral_model_autosome_ROH_freq_dir <- file.path(Neutral_model_ROH_hotspots_dir,"Gosling_plots/autosome_roh_freq")


### Inbreeding coefficient ###
Neutral_model_F_ROH_dir  <- file.path(plink_ROH_dir,"simulated/neutral_model/F_ROH")

### Expected Heterozygosity distribution ###
Neutral_model_H_e_dir <- file.path(expected_heterozygosity_dir,"simulated/neutral_model")

histogram_line_sizes <- 3
empirical_data_color <- "darkgreen"
neutral_model_color <- "blue" 

# #################################### 
# # Defining the output dirs
# #################################### 
hyperoptimizer_results_dir <- file.path(repository_path,"hyperoptimizer_results")

if (!dir.exists(hyperoptimizer_results_dir)) {
  # Create the directory
  dir.create(hyperoptimizer_results_dir, recursive = TRUE)
}


```

---
title: "`r document_title`"
subtitle: "`r document_sub`"
output:
  html_document:
    toc: true
    toc_depth: 3  
date: "`r Sys.Date()`"
editor_options: 
  markdown: 
    wrap: 72
---

## Loading libraries

```{r library()}
library(knitr)
library(ggplot2)
```

# Standard Error Confidence interval function
Confidence level <=> konfidensgrad
```{r}
# Function to calculate standard error and its confidence interval
standard_error_confidence_interval_fun <- function(observed_data, confidence_level = 0.95) {
  if ( length(observed_data) > 1 ) {
      
  # Calculate standard error
  n <- length(observed_data)
  standard_deviation <- sd(observed_data)
  standard_error <- standard_deviation / sqrt(n - 1)
  
  # Calculate confidence interval based on standard error

  alpha <- (1 - confidence_level) / 2
  margin_of_error <- qnorm(1 - alpha) * standard_error
  mean_estimate <- mean(observed_data)
  # Calculate the percentiles for the confidence interval
  confidence_interval_lower_bound <- mean_estimate - margin_of_error # 2.5th percentile (2σ)
  confidence_interval_upper_bound <- mean_estimate + margin_of_error # 97.5th percentile (2σ)
  
  # Return confidence interval
  return(c(confidence_interval_lower_bound, confidence_interval_upper_bound))

  } else {
    return(c(NA,NA)) 
    }
  
} 

```


# 1: ROH-Frequency

## 1.1 : Autosome ROH-frequencies


### 1.1.1 : Empirical - ROH frequency

```{r Empirical - ROH frequency,echo = FALSE,results= 'hide',warning = FALSE}

setwd(Empirical_data_autosome_ROH_freq_dir)

# Pattern for finding files containing "F_ROH" and ending with ".tsv"
pattern <- ".*ROH_freq.*\\.tsv$"
population_ROH_freq_files <- list.files(pattern = pattern)

#population_ROH_freq_files
# Extracting the ROH-hotspot threshold value from the suffix of the filename
parts <- unlist(strsplit(population_ROH_freq_files[1], "threshold_")) # Split the string by "threshold_"
# Extract the decimal number using regular expressions
Empirical_data_ROH_hotspot_threshold <- as.numeric(gsub("\\.tsv", "", unlist(strsplit(parts[length(parts)], "_"))[1]))

# Extract chromosome numbers from filenames
chr_numbers <- as.numeric(gsub(".*CHR(\\d+)_.*", "\\1", population_ROH_freq_files))
# cat("CHR Values:", chr_numbers, "\n")


# Sort filenames based on chromosome numbers
sorted_population_ROH_freq_files <- population_ROH_freq_files[order(chr_numbers)]

# Initialize an empty list to store the ROH-frequency data
ROH_freq_table_Empirical_Data <- list()

# Loop through each sorted .tsv file
for (file in sorted_population_ROH_freq_files) {

  file_name <- file

  ROH_freq_data <- read.table(file, header = TRUE, comment.char = "#", stringsAsFactors = FALSE, sep = "\t")
  ROH_freq_data <- ROH_freq_data[1:(nrow(ROH_freq_data)/2), ]
  # Renaming POS to POS1 and add POS2 = POS1 + 1e5
  ROH_freq_data <- transform(ROH_freq_data, POS1 = POS, POS2 = POS + 1e5 -1)
  
  # Remove the original POS column
  ROH_freq_data <- ROH_freq_data[, c("CHR", "POS1", "POS2", "COUNT", "FREQUENCY")]
  
  
  # Add chromosome_name as an attribute to the data frame
  chr_num <- as.numeric(gsub(".*CHR(\\d+)_.*", "\\1", file))
  attr(ROH_freq_data, "chromosome_name") <- chr_num
  
  # Create a list with table name and corresponding data frame
  table_info <- list(chromosome_name = chr_num, filename = file_name, data = ROH_freq_data)  # Added chromosome_name here
  
  # Append the table info to the list
  ROH_freq_table_Empirical_Data <- c(ROH_freq_table_Empirical_Data, list(table_info))
}

ROH_freq_table_Empirical_Data$ROH_hotspot_threshold <- Empirical_data_ROH_hotspot_threshold

#View(ROH_freq_table_Empirical_Data)

```

### 1.1.2: Neutral Model - ROH frequency

```{r Neutral Model - ROH frequency,echo = FALSE,results= 'hide',warning = FALSE}
setwd(Neutral_model_autosome_ROH_freq_dir)

# Pattern for finding files containing "F_ROH" and ending with """.tsv"""

pattern <- ".*ROH_freq.*\\.tsv$"
neutral_model_ROH_freq_files <- list.files(path = Neutral_model_autosome_ROH_freq_dir, pattern = pattern)

# Extract simulation numbers from the filename
sim_number_pattern <- ".*sim_(\\d+)_.*" 
sim_numbers <- as.numeric(gsub(sim_number_pattern, "\\1", neutral_model_ROH_freq_files))
# cat("CHR Values:", chr_numbers, "\n")


# Sort filenames based on chromosome numbers
sorted_neutral_model_ROH_freq_files <- neutral_model_ROH_freq_files[order(sim_numbers)]

# Initialize an empty list to store the ROH-frequency data
ROH_freq_tables_Neutral_Model <- list()

# Loop through each sorted .tsv file
for (file in sorted_neutral_model_ROH_freq_files) {

  # Extracting the ROH-hotspot threshold value from the suffix of the filename
  parts <- unlist(strsplit(file, "threshold_")) # Split the string by "threshold_"
  # Extract the decimal number using regular expressions
  simulation_ROH_hotspot_threshold <- as.numeric(gsub("\\.tsv", "", unlist(strsplit(parts[length(parts)], "_"))[1]))
  
  ROH_freq_data <- read.table(file, header = TRUE, comment.char = "#", stringsAsFactors = FALSE, sep = "\t")
  ROH_freq_data <- ROH_freq_data[1:(nrow(ROH_freq_data)/2), ]
  # Renaming POS to POS1 and add POS2 = POS1 + 1e5
  ROH_freq_data <- transform(ROH_freq_data, POS1 = POS, POS2 = POS + 1e5 -1)
  # Remove the original POS column
  ROH_freq_data <- ROH_freq_data[, c("CHR", "POS1", "POS2", "COUNT", "FREQUENCY")]
  
  # Add chromosome_name as an attribute to the data frame
  sim_num <- as.numeric(gsub(sim_number_pattern, "\\1", file))
  attr(ROH_freq_data, "Simulation_number") <- sim_num
  
  # Create a list with table name and corresponding data frame
  table_info <- list(Simulation_number = sim_num, filename = file, data = ROH_freq_data,ROH_Hotspot_threshold = simulation_ROH_hotspot_threshold)  # Added chromosome_name here
  
  # Append the table info to the list
  ROH_freq_tables_Neutral_Model <- c(ROH_freq_tables_Neutral_Model, list(table_info))
}

# ROH_freq_tables_Neutral_Model$ROH_hotspot_threshold <- neutral_model_ROH_hotspot_threshold

# Extract all ROH-hotspot thresholds from ROH_freq_tables_Neutral_Model
all_ROH_hotspot_thresholds <- unlist(lapply(ROH_freq_tables_Neutral_Model, function(table) table$ROH_Hotspot_threshold))


# Remove any non-numeric values
all_ROH_hotspot_thresholds <- all_ROH_hotspot_thresholds[!is.na(all_ROH_hotspot_thresholds) & is.numeric(all_ROH_hotspot_thresholds)]

# Check if the vector is empty and set the average accordingly
if (length(all_ROH_hotspot_thresholds) == 0) {
  Avg_ROH_hotspot_threshold <- 0
} else {
  # Calculate the overall average F_ROH
  Avg_ROH_hotspot_threshold <- mean(all_ROH_hotspot_thresholds)
}

ROH_freq_tables_Neutral_Model$Avg_ROH_hotspot_threshold <- Avg_ROH_hotspot_threshold



#View(ROH_freq_tables_Neutral_Model)
```


### 1.1.4 Summary - ROH-hotspot threshold 

```{r echo = FALSE,warning = FALSE}
cat("ROH-hotspot selection testing results://n")

# Add overAvg_ROH_hotspot_threshold from ROH_freq_tables_Neutral_Model
neutral_model_value <- round(100*ROH_freq_tables_Neutral_Model[["Avg_ROH_hotspot_threshold"]],ROH_frequency_decimals)

# Add Avg_ROH_hotspot_threshold from ROH_freq_table_Empirical_Data
empirical_model_value <- round(100*ROH_freq_table_Empirical_Data[["ROH_hotspot_threshold"]],ROH_frequency_decimals)

# Combine all values into a data frame
ROH_hotspot_threshold_values <- data.frame(
  Model = c("Neutral", "Empirical"),
  Avg_ROH_hotspot_threshold = c(neutral_model_value, empirical_model_value)
)


# Sort the data frame based on the ROH-hotspot threshold
ROH_hotspot_threshold_values_sorted <- ROH_hotspot_threshold_values[order(ROH_hotspot_threshold_values$Avg_ROH_hotspot_threshold), ]



# Print the table using knitr::kable()
knitr::kable(ROH_hotspot_threshold_values_sorted, row.names = FALSE)

```
## 1.2 ROH-hotspots - ROH Frequency and Length
```{r ,echo = FALSE,warning = FALSE}
setwd(Empirical_data_ROH_hotspots_dir)
# Pattern for finding files containing "F_ROH" and ending with ".tsv"
pattern <- "^.*ROH_Hotspot_windows*.bed$"
empirical_roh_hotspot_bed_files <- list.files(pattern = pattern)

chromosome_number_pattern <- "^german_shepherd_CHR(\\d+)_.*"
# Extract chromosome numbers
chr_numbers <- sub(chromosome_number_pattern, "\\1", empirical_roh_hotspot_bed_files)

# Sort the files based on chromosome numbers
empirical_roh_hotspot_bed_files <- empirical_roh_hotspot_bed_files[order(as.numeric(chr_numbers))]


# Create an empty list to store information for the different ROH-hotspots in
empirical_hotspot_tables <- list()

# Loop through each .bed file (ROH-hotspot allele frequency window-file)
for (file in empirical_roh_hotspot_bed_files) {
  # Extract chromosome number and window number from file name
  chromosome <- as.numeric(sub(chromosome_number_pattern, "\\1", file))
  # window <- sub(".*windows_([0-9]+)_.*", "\\1", file)
  
  # # Create table name
  # table_name <- paste("Hotspot_chr", chromosome, "_window_", window, sep = "")
  column_names <- c("CHR","POS1","POS2")
  
  # Read the .bed file into a data frame, skipping commented lines
  chromosome_bed_file_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  
  # Loop through each hotspot for the current chromosome
  for (i in 1:nrow(chromosome_bed_file_data)) {
    Hotspot_region_data <- data.frame(
    CHR = chromosome_bed_file_data[i, ][1],
    POS1 = chromosome_bed_file_data[i, ][2],
    POS2 = chromosome_bed_file_data[i, ][3] )
    
    Hotspot_length_bp <- Hotspot_region_data$POS2-Hotspot_region_data$POS1 + 1
    Hotspot_length_Mb <- Hotspot_length_bp / 1e6
    
    Hotspot_name <- paste("Hotspot_chr",chromosome, "_window_",i, sep = "")
    
      # Find the row where the variant_position_bp is within POS1 and POS2, and CHR equals chr_number
      roh_freq_windows_in_hotspot <- ROH_freq_table_Empirical_Data[[chromosome]][["data"]][ Hotspot_region_data$POS1 <= ROH_freq_table_Empirical_Data[[chromosome]][["data"]]$POS1 & ROH_freq_table_Empirical_Data[[chromosome]][["data"]]$POS2 <= Hotspot_region_data$POS2, ]
    
    avg_frequency <- mean(roh_freq_windows_in_hotspot$FREQUENCY)
    
    # Check if the selection coefficient already exists in the list
    if (!(Hotspot_name %in% names(empirical_hotspot_tables))) {
      # If it doesn't exist, create an empty data frame
      empirical_hotspot_tables[[Hotspot_name]] <- data.frame()
    }
  
     # Create a list with table name and corresponding data frame
    table_info <- list(Filename = file,Hotspot_length_bp=Hotspot_length_bp,Hotspot_length_Mb=Hotspot_length_Mb,Avg_frequency = avg_frequency, Hotspot_region_data = Hotspot_region_data) 
    
    # Append the table info to the list under selection_scenario
    empirical_hotspot_tables[[Hotspot_name]] <- c(empirical_hotspot_tables[[Hotspot_name]], table_info)
      
        }

}

# View(empirical_hotspot_tables)

```


# 2: Inbreeding coefficient

## 2.1 Empirical data

```{r echo = FALSE,warning = FALSE}
# Set the working directory to the directory containing the F_ROH files for the empirical data
setwd(Empirical_data_F_ROH_dir)

# Pattern for finding files containing "F_ROH" and ending with ".tsv"
pattern <- "^.*F_ROH.*.tsv$"
population_F_ROH_file <- list.files(pattern = pattern)

# Extract the population name from the file names
empirical_population_name <- sub("^(.*?)_F_ROH.*", "\\1", population_F_ROH_file)

# Combine population name, number, and file name
population_info <- data.frame(empirical_population_name = empirical_population_name, file_name = population_F_ROH_file)
population_info <- population_info[order(population_info$file_name), ]

# Initialize an empty list to store the data
F_ROH_table_Empirical_Data <- list()

# Loop through each .tsv file
for (i in 1:length(population_info$empirical_population_name)) {

  # Get the file name
  file <- population_info$file_name[i]

  # Read the header line
  con <- file(file, "r")
  header <- readLines(con, n = 1)
  close(con)

  # Remove "#" from the header and split it into column names
  column_names <- sub("#", "", header)
  column_names <- strsplit(column_names, "\t")[[1]]

  # Read the .tsv frequency file into a data frame
  F_ROH_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)

  # Add empirical_population_name as an attribute to the data frame
  attr(F_ROH_data, "empirical_population_name") <- population_info$empirical_population_name[i]

  # Add filename as an attribute to the data frame
  attr(F_ROH_data, "filename") <- file_name

  # Calculate the mean of the "F_ROH" column
  avg_F_ROH_population <- mean(F_ROH_data$F_ROH)
  
  # Create a list with table name and corresponding data frame
  table_info <- list(empirical_population_name = population_info$empirical_population_name[i], filename = file, data = F_ROH_data,Avg_F_ROH = avg_F_ROH_population )  # Added empirical_population_name here

  # Append the table info to the list
  F_ROH_table_Empirical_Data <- c(F_ROH_table_Empirical_Data, table_info)
}

# Creating a histogram showing the spread in F_ROH values across the empirical population 

histogram_title <- paste0("F_ROH distribution for ",empirical_population_name)
hist(F_ROH_data$F_ROH, main = histogram_title, xlab = "F_ROH", ylab = "Frequency", col = "skyblue")


empirical_avg_F_ROH_legend_text <- paste(empirical_population_name," Average: ", round(F_ROH_table_Empirical_Data[["Avg_F_ROH"]], 3))

# Add a vertical line for the average F_ROH of the empirical population
abline(v = F_ROH_table_Empirical_Data[["Avg_F_ROH"]], col = empirical_data_color, lwd = histogram_line_sizes)
# Add legend
legend("topright", legend = empirical_avg_F_ROH_legend_text,
       col = empirical_data_color, lty = 1, cex = 0.8, text.col = "black")


# Print the overall average Avg_F_ROH
cat("Overall Average Avg_F_ROH for ",empirical_population_name,":",F_ROH_table_Empirical_Data[["Avg_F_ROH"]], "//n")


# View the modified data structure
#View(F_ROH_table_Empirical_Data)

```

## 2.2 Neutral Model

```{r echo = FALSE,warning = FALSE}
# Set the working directory to the directory containing the F_ROH files for the empirical data
setwd(Neutral_model_F_ROH_dir)

# Pattern for finding files containing "F_ROH" and ending with ".tsv"
pattern <- "^.*F_ROH.*.tsv$"
simulation_F_ROH_files <- list.files(pattern = pattern)

# Extract the simulation name from the file names
sim_name <- sub("^(.*?)_F_ROH.*", "////1", simulation_F_ROH_files)

# Extract the simulation number after "sim_"
simulation_numbers <- as.integer(sub("^sim_(////d+)_.*", "////1", simulation_F_ROH_files))

# Combine simulation name, number, and file name
sim_info <- data.frame(sim_name = sim_name, simulation_number = simulation_numbers, file_name = simulation_F_ROH_files)
sim_info <- sim_info[order(sim_info$simulation_number), ]

# Initialize an empty list to store Simulated model tables
F_ROH_tables_Neutral_Model <- list()

# Loop through each .tsv file
for (i in 1:length(sim_info$sim_name)) {
 
  # Get the file name
  file <- sim_info$file_name[i]

  # Extract the table name from the file name
  file_name <- gsub("////.tsv$", "", file)
  
  # Read the header line
  con <- file(file, "r")
  header <- readLines(con, n = 1)
  close(con)
  
  # Remove "#" from the header and split it into column names
  column_names <- sub("#", "", header)
  column_names <- strsplit(column_names, "\t")[[1]]
  
  # Read the .tsv frequency file into a data frame
  F_ROH_data <- read.table(file, header = FALSE, comment.char = "#", stringsAsFactors = FALSE, col.names = column_names)
  
  # Add sim_name as an attribute to the data frame
  attr(F_ROH_data, "sim_name") <- sim_info$sim_name[i]
  
  # Create a list with table name and corresponding data frame
  table_info <- list(sim_name = sim_info$sim_name[i], filename = file_name, data = F_ROH_data)  # Added sim_name here
  
  # Append the table info to the list
  F_ROH_tables_Neutral_Model <- c(F_ROH_tables_Neutral_Model, list(table_info))
}


# Calculating Population F_ROH for each simulation (subtable)

# Loop through each table in F_ROH_tables_Neutral_Model
for (i in seq_along(F_ROH_tables_Neutral_Model)) {
  
  # Calculate the population average from the "F_ROH" column in the current simulation (subtable)
  F_ROH_tables_Neutral_Model[[i]]$Population_F_ROH <- mean(F_ROH_tables_Neutral_Model[[i]]$data$F_ROH)
  
  
}

#View(F_ROH_tables_Neutral_Model)


# Extract all population F_ROH values from F_ROH_tables_Neutral_Model
All_Population_F_ROH <- unlist(lapply(F_ROH_tables_Neutral_Model, function(table) table$Population_F_ROH))

# Remove any non-numeric values
All_Population_F_ROH <- All_Population_F_ROH[!is.na(All_Population_F_ROH) & is.numeric(All_Population_F_ROH)]

# Check if the vector is empty and set the average accordingly
if (length(All_Population_F_ROH) == 0) {
  F_ROH_tables_Neutral_Model$Estimated_Mean_Population_F_ROH <- 0
} else {
  
# Calculate the point estimate F_ROH across all 20 simulations
F_ROH_tables_Neutral_Model$Estimated_Mean_Population_F_ROH <- mean(All_Population_F_ROH)

# Calculate the bootstrap confidence interval of the point estimate F_ROH across all 20 simulations
# Store the bootstrap confidence interval in the F_ROH_tables_Neutral_Model
# F_ROH_tables_Neutral_Model$SE_CI_Estimated_Mean_Population_F_ROH <- standard_error_confidence_interval_fun(All_Population_F_ROH)

}

# Set the plot size
options(repr.plot.width = 10, repr.plot.height = 6)

# Create histogram for simulated data
hist(All_Population_F_ROH, main = "Population F_ROH for the Neutral Model simulations", xlab = "F_ROH", ylab = "Frequency", col = "skyblue")

# Determine the location of the legend based on the comparison
if (F_ROH_tables_Neutral_Model$Estimated_Mean_Population_F_ROH > F_ROH_table_Empirical_Data[["Avg_F_ROH"]]) {
  simulated_line_location <- "topright"
  empirical_line_location <- "topleft"
} else {
  simulated_line_location <- "topleft"
  empirical_line_location <- "topright"
}


# Add a red vertical line for the point estimate F_ROH across all 20 simulations
abline(v = F_ROH_tables_Neutral_Model$Estimated_Mean_Population_F_ROH, col = neutral_model_color, lwd = histogram_line_sizes)

# Add legend for the red line
legend(simulated_line_location, legend = paste("Neutral - F_ROH Point estimate: ", round(F_ROH_tables_Neutral_Model$Estimated_Mean_Population_F_ROH, 4)),
       col = neutral_model_color, lty = 1, cex = 0.8, text.col = "black")

# Add a blue vertical line for the average F_ROH of the empirical population
abline(v = F_ROH_table_Empirical_Data[["Avg_F_ROH"]], col = empirical_data_color, lwd = histogram_line_sizes)

# Add legend for the green line
legend(empirical_line_location, legend = empirical_avg_F_ROH_legend_text,
       col = empirical_data_color, lty = 1, cex = 0.8, text.col = "black", xjust = 1, yjust = 1)

# Reset par settings to default after plotting
par(mfrow = c(1, 1))



# Print the Point estimate of F_ROH across all 20 simulations
cat("Point estimate F_ROH across all 20 simulations:", F_ROH_tables_Neutral_Model$Estimated_Mean_Population_F_ROH, "//n")

print(paste("Bootstrap 95% Confidence Interval: [", 
            F_ROH_tables_Neutral_Model$SE_CI_Estimated_Mean_Population_F_ROH[1], ", ", 
            F_ROH_tables_Neutral_Model$SE_CI_Estimated_Mean_Population_F_ROH[2], "]", sep = ""))


```

## 2.4 F_ROH summary

```{r echo = FALSE,warning = FALSE}


# Extract neutral model values and CI bounds
neutral_avg_F_ROH <- F_ROH_tables_Neutral_Model[["Estimated_Mean_Population_F_ROH"]]
neutral_lower_ci <- F_ROH_tables_Neutral_Model[["SE_CI_Estimated_Mean_Population_F_ROH"]][1]
neutral_upper_ci <- F_ROH_tables_Neutral_Model[["SE_CI_Estimated_Mean_Population_F_ROH"]][2]

# Extract empirical model value
empirical_avg_F_ROH <- F_ROH_table_Empirical_Data[["Avg_F_ROH"]]
empirical_lower_ci <- NA # Placeholder for confidence interval lower bound
empirical_upper_ci <- NA # Placeholder for confidence interval upper bound

# Combine all values into a data frame
F_ROH_values <- data.frame(
  Model = c("Neutral", "Empirical"),
  F_ROH = c(neutral_avg_F_ROH, empirical_avg_F_ROH),
  Lower_CI = c(neutral_lower_ci, empirical_lower_ci),
  Upper_CI = c(neutral_upper_ci, empirical_upper_ci)
)

# Format all numeric values to have the number of decimals defined by F_ROH_values_decimals
F_ROH_values$F_ROH <- round(F_ROH_values$F_ROH,F_ROH_values_decimals)
F_ROH_values$Lower_CI <- as.numeric(round(F_ROH_values$Lower_CI,F_ROH_values_decimals))
F_ROH_values$Upper_CI <- as.numeric(round(F_ROH_values$Upper_CI,F_ROH_values_decimals))

# Sort the data frame based on F_ROH column
F_ROH_values_sorted <- F_ROH_values[order(as.numeric(F_ROH_values$F_ROH)), ]


# Print the table using knitr::kable()
knitr::kable(F_ROH_values_sorted, row.names = FALSE)

```
### 2.4.1 Visualizing F_ROH
```{r echo = FALSE,warning = FALSE}
# # Remove the empirical model from the plotting data
# plotting_data <- F_ROH_values[F_ROH_values$Model != "Empirical", ]
# 
# # Add a column indicating whether the empirical F_ROH is within the CI of each model
# empirical_F_ROH <- empirical_avg_F_ROH
# plotting_data$Empirical_Within_CI <- (plotting_data$Lower_CI <= empirical_F_ROH) & (plotting_data$Upper_CI >= empirical_F_ROH)
# 
# # Create the plot
# p <- ggplot(plotting_data, aes(x = Model, y = F_ROH, color = Empirical_Within_CI)) +
#   geom_point(size = 3) +
#   geom_errorbar(aes(ymin = Lower_CI, ymax = Upper_CI), width = 0.2) +
#   geom_hline(aes(yintercept = empirical_F_ROH), linetype = "dashed", color = "red", linewidth = 1, show.legend = TRUE) +
#   labs(title = "F_ROH Values and Confidence Intervals by Model",
#        x = "Model",
#        y = "F_ROH",
#        color = "Empirical F_ROH\nwithin CI",
#        linetype = "Empirical F_ROH") +
#   scale_linetype_manual(name = "Empirical F_ROH", values = c("dashed")) +
#   scale_color_manual(values = c("TRUE" = "blue", "FALSE" = "black"), labels = c("TRUE" = "Inside CI", "FALSE" = "Outside CI")) +
#   theme_minimal() +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
#   scale_y_continuous(limits = c(min(plotting_data$Lower_CI), max(plotting_data$Upper_CI)),
#                      breaks = round(seq(min(plotting_data$Lower_CI), max(plotting_data$Upper_CI), length.out = 10), 5))
# 
# # Print the plot
# print(p)
# 
# # Print which models the empirical F_ROH is inside of
# inside_models <- plotting_data$Model[plotting_data$Empirical_Within_CI]
# outside_models <- plotting_data$Model[!plotting_data$Empirical_Within_CI]
# 
# cat("Models where empirical F_ROH is within CI:\n")
# print(inside_models)
# cat("\nModels where empirical F_ROH is outside CI:\n")
# print(outside_models)

```



# 3: Expected Heterozygosity

## 3.1 Empirical data

```{r Expected Heterozygosity - Empirical data, echo = FALSE,warning = FALSE}
setwd(Empirical_data_H_e_dir)
# Pattern for finding the specific file containing "5th_percentiles_of_H_e_distribution.tsv"
# pattern <- "^.*5th_percentiles_of_H_e_distribution.tsv$"
pattern <- "^.*H_e_distribution_.*.tsv$"
empirical_H_e_distribution_file <- list.files(path = Empirical_data_H_e_dir, pattern = pattern)

# Check if exactly one file is found
if (length(empirical_H_e_distribution_file) != 1) {
  stop("There should be exactly one file matching the pattern.")
}

# Read the file
file <- empirical_H_e_distribution_file[1]

# Read the .tsv file into a data frame
empirical_H_e_distribution_table <- read.table(file, header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# # Set the working directory to the directory containing the H_e files for the empirical data
# setwd(Empirical_data_H_e_dir)
# 
# # List all PNG files in the directory
# # png_files <- list.files(pattern = "//.png$", full.names = TRUE)
# png_files <- list.files(pattern = "//.png$")
# 
# # Display each PNG file
# for (file in png_files) {
#   cat("File:", file, "/n")
#   # Open a PNG device
#   png::png(filename = file)
#   # Display image using readPNG() from the png package
#   raster <- png::readPNG(file)
#   graphics::plot.new()
#   graphics::plot.window(xlim = c(0, 1), ylim = c(0, 1), asp = 1)
#   graphics::rasterImage(raster, 0, 0, 1, 1)
#   # Close the PNG device
#   dev.off()
# }
# 
```

## 3.2 Neutral Model

```{r Expected Heterozygosity - Neutral Model, echo = FALSE,warning = FALSE }
setwd(Neutral_model_H_e_dir)
# Pattern for finding the specific file containing "5th_percentiles_of_H_e_distribution.tsv"
pattern <- "^.*5th_percentiles_of_H_e_distribution.tsv$"
neutral_model_5th_percentiles_of_H_e_files <- list.files(pattern = pattern)

# Check if exactly one file is found
if (length(neutral_model_5th_percentiles_of_H_e_files) != 1) {
  stop("There should be exactly one file matching the pattern.")
}

# Read the file
file <- neutral_model_5th_percentiles_of_H_e_files[1]

# Read the .tsv file into a data frame
data <- read.table(file, header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# Store the data frame in a list as a subtable called "data"
H_e_5th_percentiles_Neutral_model <- list(results = data)



# Replace NA values with 0
H_e_5th_percentiles_Neutral_model[["results"]][["Fifth_Percentile"]][is.na(H_e_5th_percentiles_Neutral_model[["results"]][["Fifth_Percentile"]])] <- 0

H_e_5th_percentiles_Neutral_model$Estimated_Mean_H_e_5th_percentile <- mean(H_e_5th_percentiles_Neutral_model[["results"]][["Fifth_Percentile"]])
H_e_5th_percentiles_Neutral_model$Estimated_Avg_Window_based_H_e <- mean(H_e_5th_percentiles_Neutral_model[["results"]][["Avg_H_e"]])

# Calculate the bootstrap confidence interval of the point estimate F_ROH across all 20 simulations
# Store the bootstrap confidence interval in the F_ROH_tables_Neutral_Model
# H_e_5th_percentiles_Neutral_model$SE_CI_Estimated_Mean_H_e_5th_percentile <- standard_error_confidence_interval_fun(H_e_5th_percentiles_Neutral_model[["results"]][["Fifth_Percentile"]])





#View(H_e_5th_percentiles_Neutral_model)


```
## 3.4 Genomewide 5th percentile comparison - Expected Heterozygosity Summary

```{r Expected Heterozygosity Summary, echo = FALSE,warning = FALSE}


# Extract neutral model values and CI bounds
neutral_avg_H_e <- H_e_5th_percentiles_Neutral_model[["Estimated_Avg_Window_based_H_e"]]
neutral_avg_H_e_5th_percentile <- H_e_5th_percentiles_Neutral_model[["Estimated_Mean_H_e_5th_percentile"]]
neutral_lower_ci <- H_e_5th_percentiles_Neutral_model[["SE_CI_Estimated_Mean_H_e_5th_percentile"]][1]
neutral_upper_ci <- H_e_5th_percentiles_Neutral_model[["SE_CI_Estimated_Mean_H_e_5th_percentile"]][2]

# Extract empirical model value
empirical_avg_H_e <- empirical_H_e_distribution_table$Avg_H_e
# empirical_avg_H_e <- NA
empirical_avg_H_e_5th_percentile <- empirical_H_e_distribution_table$Fifth_Percentile
# empirical_avg_H_e_5th_percentile <- NA
empirical_lower_ci <- NA # Placeholder for confidence interval lower bound
empirical_upper_ci <- NA # Placeholder for confidence interval upper bound

# Combine all values into a data frame
H_e_5th_percentile_values <- data.frame(
  Model = c("Neutral", "Empirical"),
  Avg_H_e = c(neutral_avg_H_e, empirical_avg_H_e),
  H_e_5th_percentile = c(neutral_avg_H_e_5th_percentile, empirical_avg_H_e_5th_percentile),
  Lower_CI = c(neutral_lower_ci, empirical_lower_ci),
  Upper_CI = c(neutral_upper_ci, empirical_upper_ci)
)

# Format all numeric values to 5 decimal places
H_e_5th_percentile_values$Avg_H_e <- round(H_e_5th_percentile_values$Avg_H_e, H_e_values_decimals)
H_e_5th_percentile_values$H_e_5th_percentile <- round(H_e_5th_percentile_values$H_e_5th_percentile, H_e_values_decimals)
H_e_5th_percentile_values$Lower_CI <- round(H_e_5th_percentile_values$Lower_CI,H_e_values_decimals)
H_e_5th_percentile_values$Upper_CI <- round(H_e_5th_percentile_values$Upper_CI,H_e_values_decimals)

# Sort the data frame based on H_e_5th_percentile column
H_e_5th_percentile_values_sorted <- H_e_5th_percentile_values[order(as.numeric(H_e_5th_percentile_values$H_e_5th_percentile)), ]


# Print the table using knitr::kable()
knitr::kable(H_e_5th_percentile_values_sorted, row.names = FALSE)

```




# 4: Summary

## 4.0 General comparison
### 4.0.1 ROH-hotspot threshold comparison
```{r echo = FALSE,warning = FALSE}

cat("\n ROH-hotspot threshold comparison between the different datasets")
# Print the table using knitr::kable()
knitr::kable(ROH_hotspot_threshold_values_sorted, row.names = FALSE)
```
### 4.0.2 F_ROH comparison
```{r echo = FALSE,warning = FALSE}
 # Print the table using knitr::kable()
knitr::kable(F_ROH_values_sorted, row.names = FALSE)
```

# Cost function computation

## Extracting the simulation values

```{r}
reference_H_e_5th_percentile_No_MAF <- H_e_5th_percentile_values_sorted$H_e_5th_percentile[H_e_5th_percentile_values_sorted$Model=="Empirical"]
# reference_H_e_5th_percentile_No_MAF <- 0.0742
neutral_model_5th_percentile_No_MAF <- H_e_5th_percentile_values_sorted$H_e_5th_percentile[H_e_5th_percentile_values_sorted$Model=="Neutral"]
# neutral_model_5th_percentile_No_MAF <- 0.07385

reference_Avg_H_e_No_MAF <- H_e_5th_percentile_values_sorted$Avg_H_e[H_e_5th_percentile_values_sorted$Model=="Empirical"]
# reference_Avg_H_e_No_MAF <- 0.244588241178689
neutral_model_Avg_H_e_No_MAF <- H_e_5th_percentile_values_sorted$Avg_H_e[H_e_5th_percentile_values_sorted$Model=="Neutral"] 

reference_F_ROH <- F_ROH_values_sorted$F_ROH[F_ROH_values_sorted$Model=="Empirical"]
# reference_F_ROH <- 0.233
neutral_model_F_ROH <- F_ROH_values_sorted$F_ROH[F_ROH_values_sorted$Model=="Neutral"]
# neutral_model_F_ROH <- 0.158


reference_ROH_hotspot_threshold <- ROH_hotspot_threshold_values_sorted$Avg_ROH_hotspot_threshold[ROH_hotspot_threshold_values_sorted$Model=="Empirical"]


reference_ROH_hotspot_threshold <- ifelse(reference_ROH_hotspot_threshold == 0,

                                                 0,

                                                reference_ROH_hotspot_threshold / 100)

# reference_ROH_hotspot_threshold <- 0.554

neutral_model_ROH_hotspot_threshold <- ROH_hotspot_threshold_values_sorted$Avg_ROH_hotspot_threshold[ROH_hotspot_threshold_values_sorted$Model=="Neutral"]

neutral_model_ROH_hotspot_threshold <- ifelse(neutral_model_ROH_hotspot_threshold == 0,
                                                 0,
                                                neutral_model_ROH_hotspot_threshold / 100)

# neutral_model_ROH_hotspot_threshold <- 0.313

```
## Compute absolute difference to empirical data

```{r}
H_e_5th_percentile_diff <- reference_H_e_5th_percentile_No_MAF- neutral_model_5th_percentile_No_MAF
cat("\n\nEmpirical H_e 5th % - Neutral Model H_e 5th% = \n ",reference_H_e_5th_percentile_No_MAF,"-",neutral_model_5th_percentile_No_MAF,"=",H_e_5th_percentile_diff)

Avg_H_e_diff <- reference_Avg_H_e_No_MAF- neutral_model_Avg_H_e_No_MAF
cat("\n\nEmpirical Avg H_e - Neutral Model Avg H_e = \n ",reference_Avg_H_e_No_MAF,"-",neutral_model_Avg_H_e_No_MAF,"=",Avg_H_e_diff)

F_ROH_diff <- reference_F_ROH - neutral_model_F_ROH
cat("\n\nEmpirical F_ROH - Neutral Model F_ROH = \n",reference_F_ROH,"-",neutral_model_F_ROH,"=",F_ROH_diff)

ROH_hotspot_threshold_diff <- reference_ROH_hotspot_threshold - neutral_model_ROH_hotspot_threshold
cat("\n\nEmpirical ROH-Hotspot threshold - Neutral Model ROH-Hotspot threshold = \n ",reference_ROH_hotspot_threshold,"-",neutral_model_ROH_hotspot_threshold,"=",ROH_hotspot_threshold_diff)

```
# Cost function computation

```{r Cost function}
# Calculate percentage deviation for H_e_5th_percentile
percentage_deviation_H_e_5th_percentile <- ifelse(H_e_5th_percentile_diff == 0, 
                                                 0, 
                                                 abs(H_e_5th_percentile_diff) / reference_H_e_5th_percentile_No_MAF)

# Calculate percentage deviation for Average H_e
percentage_deviation_Avg_H_e <- ifelse(Avg_H_e_diff == 0, 
                                                 0, 
                                                 abs(Avg_H_e_diff) / reference_Avg_H_e_No_MAF)

# Calculate percentage deviation for F_ROH
percentage_deviation_F_ROH <- ifelse(F_ROH_diff == 0, 
                                     0, 
                                     abs(F_ROH_diff) / reference_F_ROH)

# Calculate percentage deviation for ROH_hotspot_threshold
percentage_deviation_ROH_hotspot_threshold <- ifelse(ROH_hotspot_threshold_diff == 0, 
                                                    0, 
                                                    abs(ROH_hotspot_threshold_diff) / reference_ROH_hotspot_threshold)

H_e_weight <- 2
```

# Cost function with H_e_5th_percentile
```{r Cost function}
H_e_weight <- 2

# Output the cost (percentage deviation)
cost_old <- H_e_weight*percentage_deviation_H_e_5th_percentile + percentage_deviation_F_ROH + percentage_deviation_ROH_hotspot_threshold
cost_old <- round(cost_old,decimals_output_values)
cat("\n Cost=",H_e_weight,"*",percentage_deviation_H_e_5th_percentile,"+",percentage_deviation_F_ROH,"+",percentage_deviation_ROH_hotspot_threshold,"=\n",cost_old)
```
# Cost function with H_e
```{r}
# Output the cost (percentage deviation)
cost <- H_e_weight*percentage_deviation_Avg_H_e + percentage_deviation_F_ROH + percentage_deviation_ROH_hotspot_threshold
cost <- round(cost,decimals_output_values)
cat("\n Cost=",H_e_weight,"*",percentage_deviation_Avg_H_e,"+",percentage_deviation_F_ROH,"+",percentage_deviation_ROH_hotspot_threshold,"=\n",cost)

```

# Exporting the results
```{r Export Results}


# Define the filename with the output directory path
simulation_results_file <- file.path(hyperoptimizer_results_dir, paste("neutral_models_cost_function_results",".tsv", sep = ""))

# # Create an empty data frame if the file does not exist
# column_names <- c("Chr","NeBurnIn","nInd","InbredFo","NeBottleneck","nGenBottleneck","nGenBreed","nBreed","SNPchipRefPop","Sim_H_e_5th_perc","Sim_F_ROH","Sim_ROH_hotspot_thr","Sim_Cost_Result")
# 
# 
# # Create a list of the values for the current simulation
# current_simulation_values <- list(chr,Ne_burn_in,nInd_fo,Inbred_fo,N_e_bottleneck,n_generations_bottleneck,n_gen_breed,n_breed,snp_chip,
#          round(neutral_model_5th_percentile_No_MAF,decimals_output_values),
#          round(neutral_model_F_ROH,decimals_output_values),
#          round(neutral_model_ROH_hotspot_threshold,decimals_output_values),
#          round(cost,decimals_output_values))

# # Create an empty data frame if the file does not exist
# column_names <- c("Chr","NeBurnIn","nInd","InbredFo","NeBottleneck","nGenBottleneck","nGenBreed","nBreed","SNPchipRefPop","Sim_H_e","Sim_F_ROH","Sim_ROH_hotspot_thr","Sim_Cost_Result")
# 
# 
# # Create a list of the values for the current simulation
# current_simulation_values <- list(chr,Ne_burn_in,nInd_fo,Inbred_fo,N_e_bottleneck,n_generations_bottleneck,n_gen_breed,n_breed,snp_chip,
#          round(neutral_model_Avg_H_e_No_MAF,decimals_output_values),
#          round(neutral_model_F_ROH,decimals_output_values),
#          round(neutral_model_ROH_hotspot_threshold,decimals_output_values),
#          round(cost,decimals_output_values))

# Create an empty data frame if the file does not exist
column_names <- c("Chr","NeBurnIn","nInd","InbredFo","NeBottleneck","nGenBottleneck","nGenBreed","nBreed","SNPchipRefPop","Mutations","Sim_H_e","Sim_H_e_5th_perc","Sim_F_ROH","Sim_ROH_hotspot_thr","Sim_Cost_Result")

# Create a list of the values for the current simulation
current_simulation_values <- list(chr,Ne_burn_in,nInd_fo,Inbred_fo,N_e_bottleneck,n_generations_bottleneck,n_gen_breed,n_breed,snp_chip,Introduce_mutations,
         round(neutral_model_Avg_H_e_No_MAF,decimals_output_values),
         round(neutral_model_5th_percentile_No_MAF,decimals_output_values),
         round(neutral_model_F_ROH,decimals_output_values),
         round(neutral_model_ROH_hotspot_threshold,decimals_output_values),
         round(cost,decimals_output_values))

# Convert the list to a data frame
current_simulation_df <- data.frame(matrix(unlist(current_simulation_values), nrow=1, byrow=TRUE))
names(current_simulation_df) <- column_names

# Check if the file exists
if (file.exists(simulation_results_file)) {
    # Read the existing data from the file
    all_simulation_results_df <- read.table(simulation_results_file, header = TRUE, sep = "\t")
    
    # Append the new row to the existing data frame
    all_simulation_results_df <- rbind(all_simulation_results_df, current_simulation_df)
} else {
    # Create an empty data frame if the file does not exist
    all_simulation_results_df <- current_simulation_df
}

# Write data to TSV file without quotes and with tab separation
write.table(all_simulation_results_df, file = simulation_results_file, sep = "\t", row.names = FALSE, quote = FALSE)
```

